# <h1 align=center>Toranto Data</h1>

by: Nishad Thakur

Notebook has 3 parts separated by 3 questions

### 1) Import DATA - Web scraping

In [1]:
!pip install bs4
import pandas as pd
import requests
from bs4 import BeautifulSoup

  Created wheel for bs4: filename=bs4-0.0.1-py3-none-any.whl size=1273 sha256=be5f6928dcc6df7be847f7da4c11818a92d198a6fe851bbd42de566b4ffa84b0
  Stored in directory: /tmp/wsuser/.cache/pip/wheels/0a/9e/ba/20e5bbc1afef3a491f0b3bb74d508f99403aabe76eda2167ca
Successfully built bs4


In [2]:
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
wiki_data=requests.get(url).text
soup = BeautifulSoup(wiki_data, 'html5lib') #soup for getting data
table_contents=[]
table=soup.find('table')
len(table)

2

In [3]:
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

print(table_contents)   #since its not vertical table we need to clean data

[{'PostalCode': 'M3A', 'Borough': 'North York', 'Neighborhood': 'Parkwoods'}, {'PostalCode': 'M4A', 'Borough': 'North York', 'Neighborhood': 'Victoria Village'}, {'PostalCode': 'M5A', 'Borough': 'Downtown Toronto', 'Neighborhood': 'Regent Park, Harbourfront'}, {'PostalCode': 'M6A', 'Borough': 'North York', 'Neighborhood': 'Lawrence Manor, Lawrence Heights'}, {'PostalCode': 'M7A', 'Borough': "Queen's Park", 'Neighborhood': 'Ontario Provincial Government'}, {'PostalCode': 'M9A', 'Borough': 'Etobicoke', 'Neighborhood': 'Islington Avenue'}, {'PostalCode': 'M1B', 'Borough': 'Scarborough', 'Neighborhood': 'Malvern, Rouge'}, {'PostalCode': 'M3B', 'Borough': 'North York', 'Neighborhood': 'Don Mills North'}, {'PostalCode': 'M4B', 'Borough': 'East York', 'Neighborhood': 'Parkview Hill, Woodbine Gardens'}, {'PostalCode': 'M5B', 'Borough': 'Downtown Toronto', 'Neighborhood': 'Garden District, Ryerson'}, {'PostalCode': 'M6B', 'Borough': 'North York', 'Neighborhood': 'Glencairn'}, {'PostalCode': 'M9

In [4]:
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})
df.head()

PostalCode           Borough                      Neighborhood
0        M3A        North York                         Parkwoods
1        M4A        North York                  Victoria Village
2        M5A  Downtown Toronto         Regent Park, Harbourfront
3        M6A        North York  Lawrence Manor, Lawrence Heights
4        M7A      Queen's Park     Ontario Provincial Government

In [5]:
df_postcode = df.groupby(['PostalCode', 'Borough'])['Neighborhood'].apply(', '.join).reset_index()
df_postcode.head()


PostalCode      Borough                            Neighborhood
0        M1B  Scarborough                          Malvern, Rouge
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

In [6]:
# Shape of data frame
df_postcode.shape

(103, 3)

Answer of question 1 is (103,3)

## latitudes and longitudes for various neighbourhoods in Canada

In [9]:
!pip install geocoder  #Geocoder for lattitude and longitude data

     |████████████████████████████████| 98 kB 8.1 MB/s  eta 0:00:01


In [10]:
import geocoder # import geocoder

In [11]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [12]:
print("The shape of our wiki data is: ", df_postcode.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


In [15]:
df_postcode.dtypes

PostalCode      object
Borough         object
Neighborhood    object
dtype: object

In [14]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [18]:
combined_data = df_postcode.join(data.set_index('Postal Code'), on='PostalCode', how='inner')
combined_data

PostalCode              Borough  \
0          M1B          Scarborough   
1          M1C          Scarborough   
2          M1E          Scarborough   
3          M1G          Scarborough   
4          M1H          Scarborough   
..         ...                  ...   
98         M9N                 York   
99         M9P            Etobicoke   
100        M9R            Etobicoke   
101        M9V            Etobicoke   
102        M9W  Etobicoke Northwest   

                                          Neighborhood   Latitude  Longitude  
0                                       Malvern, Rouge  43.806686 -79.194353  
1               Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497  
2                    Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.239476  
..                                                 ...        ...        ...  
98                                              Weston  43.706876 -79.518188  
99                                           Westmount  43.696319 -79.532242  
100  Kingsview Village, St. Phillips, Martin Grove ...  43.688905 -79.554724  
101  South Steeles, Silverstone, Humbergate, Jamest...  43.739416 -79.588437  
102  Clairville, Humberwood, Woodbine Downs, West H...  43.706748 -79.594054  

[103 rows x 5 columns]

Question 2 we got 103 rows with geo data

## 3) Cluster

Drawing inspiration from the previous lab where we cluster the neighbourhood of NYC, We cluster Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [19]:
from geopy.geocoders import Nominatim

In [20]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


In [21]:
import folium #visualization

In [24]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighborhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

In [25]:
CLIENT_ID = 'O1LW2JUKJUJYE2ODTTVL2UPKT0URN45ZPDCHKIS0KY3OGMDC' 
CLIENT_SECRET = 'RMTYY4NRDXGY04EQNZFI2IQ0SFNICO11XFLPYJB4OJQDLZTU'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)  

Your credentails:
CLIENT_ID: O1LW2JUKJUJYE2ODTTVL2UPKT0URN45ZPDCHKIS0KY3OGMDC
CLIENT_SECRET:RMTYY4NRDXGY04EQNZFI2IQ0SFNICO11XFLPYJB4OJQDLZTU


In [28]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

In [29]:
venues_in_toronto = getNearbyVenues(combined_data['Neighborhood'], combined_data['Latitude'], combined_data['Longitude'])

Malvern, Rouge
Rouge Hill, Port Union, Highland Creek
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
Kennedy Park, Ionview, East Birchmount Park
Golden Mile, Clairlea, Oakridge
Cliffside, Cliffcrest, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Wexford Heights, Scarborough Town Centre
Wexford, Maryvale
Agincourt
Clarks Corners, Tam O'Shanter, Sullivan
Milliken, Agincourt North, Steeles East, L'Amoreaux East
Steeles West, L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
York Mills, Silver Hills
Willowdale, Newtonbrook
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Don Mills South
Bathurst Manor, Wilson Heights, Downsview North
Northwood Park, York University
Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Parkview Hill, Woodbine Gardens
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
The Danforth  East
The Danforth West, Riverdale


In [30]:
venues_in_toronto.shape

(1332, 5)

So we have 1332 records and 5 columns. Checking sample data

In [31]:
venues_in_toronto.head()

Neighborhood  Neighbourhood Latitude  \
0                          Malvern, Rouge               43.806686   
1  Rouge Hill, Port Union, Highland Creek               43.784535   
2  Rouge Hill, Port Union, Highland Creek               43.784535   
3  Rouge Hill, Port Union, Highland Creek               43.784535   
4       Guildwood, Morningside, West Hill               43.763573   

   Neighbourhood Longitude                                              Venue  \
0               -79.194353                                            Wendy’s   
1               -79.160497                             Chris Effects Painting   
2               -79.160497                              Royal Canadian Legion   
3               -79.160497  SEBS Engineering Inc. (Sustainable Energy and ...   
4               -79.188711                                     RBC Royal Bank   

               Venue Category  
0        Fast Food Restaurant  
1  Construction & Landscaping  
2                         Bar  
3  Construction & Landscaping  
4                        Bank

Checking the Venues based on Neighbourhood

In [34]:
venues_in_toronto.groupby('Neighborhood').head()

Neighborhood  \
0                                        Malvern, Rouge   
1                Rouge Hill, Port Union, Highland Creek   
2                Rouge Hill, Port Union, Highland Creek   
3                Rouge Hill, Port Union, Highland Creek   
4                     Guildwood, Morningside, West Hill   
...                                                 ...   
1327  Clairville, Humberwood, Woodbine Downs, West H...   
1328  Clairville, Humberwood, Woodbine Downs, West H...   
1329  Clairville, Humberwood, Woodbine Downs, West H...   
1330  Clairville, Humberwood, Woodbine Downs, West H...   
1331  Clairville, Humberwood, Woodbine Downs, West H...   

      Neighbourhood Latitude  Neighbourhood Longitude  \
0                  43.806686               -79.194353   
1                  43.784535               -79.160497   
2                  43.784535               -79.160497   
3                  43.784535               -79.160497   
4                  43.763573               -79.188711   
...                      ...                      ...   
1327               43.706748               -79.594054   
1328               43.706748               -79.594054   
1329               43.706748               -79.594054   
1330               43.706748               -79.594054   
1331               43.706748               -79.594054   

                                                  Venue  \
0                                               Wendy’s   
1                                Chris Effects Painting   
2                                 Royal Canadian Legion   
3     SEBS Engineering Inc. (Sustainable Energy and ...   
4                                        RBC Royal Bank   
...                                                 ...   
1327                                 Economy Rent A Car   
1328                             Logistics Distribution   
1329                                      Saand Rexdale   
1330                                          PC Garden   
1331                               Vectra Heavy Haulers   

                  Venue Category  
0           Fast Food Restaurant  
1     Construction & Landscaping  
2                            Bar  
3     Construction & Landscaping  
4                           Bank  
...                          ...  
1327         Rental Car Location  
1328                         Bar  
1329                   Drugstore  
1330               Garden Center  
1331                  Truck Stop  

[433 rows x 5 columns]

So there are 433 records for each neighbourhood.

In [35]:
venues_in_toronto.groupby('Venue Category').max()

Neighborhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Airport                                                Downsview East   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
Airport Service     CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                     Toronto Dominion Centre, Design Exchange   
Wings Joint                             Ontario Provincial Government   
Women's Store                                     Caledonia-Fairbanks   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.718518               -79.464763   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Lounge                   43.628947               -79.394420   
Airport Service                  43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.653206               -79.379817   
Wings Joint                      43.662301               -79.389494   
Women's Store                    43.689026               -79.453512   
Yoga Studio                      43.715383               -79.321558   

                                              Venue  
Venue Category                                       
Accessories Store               Ardene Shoes Outlet  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Lounge                        Porter Lounge  
Airport Service            ORNGE - Toronto Air Base  
...                                             ...  
Warehouse Store                              Costco  
Wine Bar                          The National Club  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[234 rows x 4 columns]

There are around 234 different types of Venue Categories

In [36]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Airport  Airport Food Court  Airport Lounge  \
0                     0        0                   0               0   
1                     0        0                   0               0   
2                     0        0                   0               0   
3                     0        0                   0               0   
4                     0        0                   0               0   
...                 ...      ...                 ...             ...   
1327                  0        0                   0               0   
1328                  0        0                   0               0   
1329                  0        0                   0               0   
1330                  0        0                   0               0   
1331                  0        0                   0               0   

      Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                   0                 0                    0             0   
1                   0                 0                    0             0   
2                   0                 0                    0             0   
3                   0                 0                    0             0   
4                   0                 0                    0             0   
...               ...               ...                  ...           ...   
1327                0                 0                    0             0   
1328                0                 0                    0             0   
1329                0                 0                    0             0   
1330                0                 0                    0             0   
1331                0                 0                    0             0   

      Aquarium  Art Gallery  ...  Truck Stop  Vegetarian / Vegan Restaurant  \
0            0            0  ...           0                              0   
1            0            0  ...           0                              0   
2            0            0  ...           0                              0   
3            0            0  ...           0                              0   
4            0            0  ...           0                              0   
...        ...          ...  ...         ...                            ...   
1327         0            0  ...           0                              0   
1328         0            0  ...           0                              0   
1329         0            0  ...           0                              0   
1330         0            0  ...           0                              0   
1331         0            0  ...           1                              0   

      Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                    0            0                      0                0   
1                    0            0                      0                0   
2                    0            0                      0                0   
3                    0            0                      0                0   
4                    0            0                      0                0   
...                ...          ...                    ...              ...   
1327                 0            0                      0                0   
1328                 0            0                      0                0   
1329                 0            0                      0                0   
1330                 0            0                      0                0   
1331                 0            0                      0                0   

      Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0            0            0              0            0  
1            0            0              0            0  
2            0            0              0            0  
3            0            0              0            0  
4            0  

Adding the neighbourhood to the encoded dataframe

In [38]:
toronto_venue_cat['Neighborhood'] = venues_in_toronto['Neighborhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Yoga Studio  Accessories Store  Airport  Airport Food Court  \
0            0                  0        0                   0   
1            0                  0        0                   0   
2            0                  0        0                   0   
3            0                  0        0                   0   
4            0                  0        0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  ...  Train Station  Truck Stop  \
0             0         0  ...              0           0   
1             0         0  ...              0           0   
2             0         0  ...              0           0   
3             0         0  ...              0           0   
4             0         0  ...              0           0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  
0              0  
1              0  
2              0  
3              0  
4              0  

[5 rows x 234 columns]

In [39]:
toronto_grouped = toronto_venue_cat.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Yoga Studio  \
0                                        Agincourt          0.0   
1                           Alderwood, Long Branch          0.0   
2  Bathurst Manor, Wilson Heights, Downsview North          0.0   
3                                  Bayview Village          0.0   
4                Bedford Park, Lawrence Manor East          0.0   

   Accessories Store  Airport  Airport Food Court  Airport Lounge  \
0                0.0      0.0                 0.0             0.0   
1                0.0      0.0                 0.0             0.0   
2                0.0      0.0                 0.0             0.0   
3                0.0      0.0                 0.0             0.0   
4                0.0      0.0                 0.0             0.0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  ...  \
0              0.0               0.0             0.000000           0.0  ...   
1              0.0               0.0             0.000000           0.0  ...   
2              0.0               0.0             0.000000           0.0  ...   
3              0.0               0.0             0.000000           0.0  ...   
4              0.0               0.0             0.043478           0.0  ...   

   Train Station  Truck Stop  Vegetarian / Vegan Restaurant  Video Game Store  \
0            0.0         0.0                            0.0               0.0   
1            0.0         0.0                            0.0               0.0   
2            0.0         0.0                            0.0               0.0   
3            0.0         0.0                            0.0               0.0   
4            0.0         0.0                            0.0               0.0   

   Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0          0.0                    0.0              0.0       0.0          0.0   
1          0.0                    0.0              0.0       0.0          0.0   
2          0.0                    0.0              0.0       0.0          0.0   
3          0.0                    0.0              0.0       0.0          0.0   
4          0.0                    0.0              0.0       0.0          0.0   

   Women's Store  
0            0.0  
1            0.0  
2            0.0  
3            0.0  
4            0.0  

[5 rows x 234 columns]

In [40]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [41]:
import numpy as np #lets play with numbers

Top 10 venues

In [42]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village   Japanese Restaurant   
4                Bedford Park, Lawrence Manor East    Italian Restaurant   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot  Latin American Restaurant          Skating Rink   
1            Playground                        Pub          Skating Rink   
2           Coffee Shop                Pizza Place      Sushi Restaurant   
3                  Bank                       Café    Chinese Restaurant   
4        Sandwich Place                Coffee Shop           Pizza Place   

  5th Most Common Venue     6th Most Common Venue     7th Most Common Venue  \
0           Yoga Studio            Massage Studio  Mediterranean Restaurant   
1           Coffee Shop                  Pharmacy                       Gym   
2    Frozen Yogurt Shop            Sandwich Place               Supermarket   
3           Yoga Studio  Mediterranean Restaurant               Men's Store   
4                   Spa                   Butcher                      Café   

  8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
0           Men's Store         Metro Station         Mexican Restaurant  
1        Sandwich Place        Medical Center   Mediterranean Restaurant  
2           Bridal Shop         Shopping Mall                Gas Station  
3         Metro Station    Mexican Restaurant  Middle Eastern Restaurant  
4          Liquor Store      Sushi Restaurant                  Juice Bar

In [43]:
#time to make some clusters
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [44]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

Checking the labelling of our model

In [45]:
kmeans.labels_[0:100]

array([4, 0, 4, 4, 1, 4, 4, 1, 4, 2, 4, 1, 4, 4, 4, 0, 1, 4, 4, 4, 4, 4,
       1, 4, 4, 2, 1, 4, 4, 1, 4, 4, 4, 1, 1, 4, 4, 0, 4, 4, 4, 4, 4, 0,
       4, 4, 4, 1, 4, 0, 1, 2, 1, 4, 3, 4, 4, 2, 4, 4, 1, 1, 4, 1, 4, 0,
       2, 1, 1, 2, 4, 4, 4, 0, 4, 4, 4, 4, 4, 1, 1, 4, 4, 2, 4, 2, 4, 1,
       4, 1, 0, 4, 4, 4, 0, 2, 1, 4, 2], dtype=int32)

Let's add the clustering Label column to the top 10 common venue categories

In [46]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting

In [47]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head()

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Malvern, Rouge  43.806686   
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels       1st Most Common Venue  \
0 -79.194353             3.0        Fast Food Restaurant   
1 -79.160497             4.0  Construction & Landscaping   
2 -79.188711             4.0              Medical Center   
3 -79.216917             1.0                 Coffee Shop   
4 -79.239476             4.0         Fried Chicken Joint   

  2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
0           Yoga Studio                  Motel        Massage Studio   
1                   Bar            Yoga Studio         Movie Theater   
2   Rental Car Location             Donut Shop                  Bank   
3          Soccer Field  Korean BBQ Restaurant           Yoga Studio   
4  Caribbean Restaurant                 Bakery           Gas Station   

  5th Most Common Venue     6th Most Common Venue     7th Most Common Venue  \
0        Medical Center  Mediterranean Restaurant               Men's Store   
1        Medical Center  Mediterranean Restaurant               Men's Store   
2          Intersection                Restaurant         Electronics Store   
3                 Motel            Medical Center  Mediterranean Restaurant   
4                  Bank        Athletics & Sports           Thai Restaurant   

  8th Most Common Venue      9th Most Common Venue     10th Most Common Venue  
0         Metro Station         Mexican Restaurant  Middle Eastern Restaurant  
1         Metro Station         Mexican Restaurant  Middle Eastern Restaurant  
2    Mexican Restaurant             Breakfast Spot         Miscellaneous Shop  
3           Men's Store              Metro Station         Mexican Restaurant  
4      Hakka Restaurant  Middle Eastern Restaurant         Miscellaneous Shop

let's remove null values

In [48]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

some action Plotting the clusters on the map

In [49]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [50]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighborhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Let's verify each of our clusters

Cluster 1

In [51]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
13   Scarborough             0.0           Pizza Place  Fast Food Restaurant   
24    North York             0.0           Coffee Shop         Grocery Store   
35     East York             0.0           Pizza Place  Gym / Fitness Center   
72    North York             0.0           Pizza Place      Sushi Restaurant   
81          York             0.0     Convenience Store         Grocery Store   
89     Etobicoke             0.0           Pizza Place            Playground   
96    North York             0.0           Pizza Place           Yoga Studio   
99     Etobicoke             0.0           Pizza Place           Coffee Shop   
100    Etobicoke             0.0        Sandwich Place           Pizza Place   

         3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
13          Chinese Restaurant              Pharmacy       Thai Restaurant   
24                 Pizza Place              Pharmacy           Supermarket   
35                Intersection           Flea Market                  Café   
72                      Bakery   Japanese Restaurant           Yoga Studio   
81                 Pizza Place              Bus Line           Yoga Studio   
89                         Pub          Skating Rink           Coffee Shop   
96               Movie Theater        Massage Studio        Medical Center   
99   Middle Eastern Restaurant          Intersection    Chinese Restaurant   
100          Mobile Phone Shop              Bus Line         Movie Theater   

        6th Most Common Venue       7th Most Common Venue  \
13        Fried Chicken Joint          Italian Restaurant   
24         Mexican Restaurant  Modern European Restaurant   
35                       Bank          Athletics & Sports   
72        Monument / Landmark              Massage Studio   
81                      Motel              Medical Center   
89                   Pharmacy                         Gym   
96   Mediterranean Restaurant                 Men's Store   
99             Sandwich Place              Discount Store   
100            Medical Center    Mediterranean Restaurant   

        8th Most Common Venue     9th Most Common Venue  \
13                Gas Station              Noodle House   
24          Mobile Phone Shop        Miscellaneous Shop   
35                  Gastropub            Breakfast Spot   
72             Medical Center  Mediterranean Restaurant   
81   Mediterranean Restaurant               Men's Store   
89             Sandwich Place            Medical Center   
96              Metro Station        Mexican Restaurant   
99         Miscellaneous Shop        Mexican Restaurant   
100               Men's Store             Metro Station   

        10th Most Common Venue  
13                        Bank  
24   Middle Eastern Restaurant  
35                    Pharmacy  
72                 Men's Store  
81               Metro Station  
89    Mediterranean Restaurant  
96   Middle Eastern Restaurant  
99           Mobile Phone Shop  
100         Mexican Restaurant

Cluster 2

In [52]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
3        Scarborough             1.0           Coffee Shop   
6        Scarborough             1.0           Coffee Shop   
8        Scarborough             1.0                 Motel   
18        North York             1.0        Clothing Store   
27        North York             1.0           Coffee Shop   
29        North York             1.0  Caribbean Restaurant   
33        North York             1.0          Liquor Store   
34        North York             1.0          Intersection   
38         East York             1.0           Coffee Shop   
43      East Toronto             1.0           Coffee Shop   
46   Central Toronto             1.0        Clothing Store   
49   Central Toronto             1.0           Coffee Shop   
53  Downtown Toronto             1.0           Coffee Shop   
57  Downtown Toronto             1.0           Coffee Shop   
58  Downtown Toronto             1.0           Coffee Shop   
60  Downtown Toronto             1.0           Coffee Shop   
62        North York             1.0    Italian Restaurant   
70  Downtown Toronto             1.0                  Café   
71        North York             1.0        Clothing Store   
78      West Toronto             1.0                  Café   
85      Queen's Park             1.0           Coffee Shop   
86       Mississauga             1.0           Coffee Shop   

     2nd Most Common Venue  3rd Most Common Venue   4th Most Common Venue  \
3             Soccer Field  Korean BBQ Restaurant             Yoga Studio   
6        Convenience Store             Hobby Shop        Department Store   
8      American Restaurant           Skating Rink             Yoga Studio   
18             Coffee Shop              Juice Bar                    Bank   
27                     Gym             Restaurant                Bus Line   
29             Coffee Shop     Falafel Restaurant                     Bar   
33           Grocery Store     Athletics & Sports             Coffee Shop   
34   Portuguese Restaurant            Coffee Shop            Hockey Arena   
38     Sporting Goods Shop           Burger Joint  Furniture / Home Store   
43                  Bakery                   Café     American Restaurant   
46             Coffee Shop            Yoga Studio                    Café   
49            Liquor Store                    Pub              Bagel Shop   
53                    Park                 Bakery          Breakfast Spot   
57      Italian Restaurant                   Café             Yoga Studio   
58                    Café   Gym / Fitness Center             Record Shop   
60                    Café    Japanese Restaurant              Restaurant   
62          Sandwich Place            Coffee Shop             Pizza Place   
70             Coffee Shop             Restaurant      Seafood Restaurant   
71  Furniture / Home Store  Vietnamese Restaurant             Coffee Shop   
78             Coffee Shop         Breakfast Spot            Intersection   
85        Sushi Restaurant            Yoga Studio              Restaurant   
86                   Hotel           Intersection     Fried Chicken Joint   

     5th Most Common Venue      6th Most Common Venue  \
3                    Motel             Medical Center   
6           Discount Store          Mobile Phone Shop   
8           Massage Studio             Medical Center   
18              Restaurant        Japanese Restaurant   
27        Asian Restaurant         Chinese Restaurant   
29  Furniture / Home Store             Massage Studio   
33             Yoga Studio         Miscellaneous Shop   
34                  Museum   Mediterranean Restaurant   
38                    Bank             Breakfast Spot   
43             Yoga Studio                    Gay Bar   
46     Sporting Goods Shop                        Spa   
49     Fried Chicken Joint           Sushi Restaurant   
53                    Café          French Restaurant   
57        Ramen 

Cluster 3

In [53]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
21              North York             2.0                  Park   
23              North York             2.0     Convenience Store   
25              North York             2.0     Food & Drink Shop   
30              North York             2.0               Airport   
40  East York/East Toronto             2.0                  Park   
44         Central Toronto             2.0                  Park   
48         Central Toronto             2.0                  Park   
50        Downtown Toronto             2.0                  Park   
74                    York             2.0                  Park   
90               Etobicoke             2.0                  Pool   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
21           Yoga Studio         Movie Theater        Massage Studio   
23                  Park           Yoga Studio                 Motel   
25                  Park           Yoga Studio                 Motel   
30           Snack Place                  Park           Yoga Studio   
40     Convenience Store          Intersection           Yoga Studio   
44           Swim School              Bus Line           Yoga Studio   
48            Restaurant          Tennis Court           Yoga Studio   
50            Playground                 Trail                 Motel   
74         Women's Store                  Pool        Hardware Store   
90                  Park                 River           Yoga Studio   

   5th Most Common Venue     6th Most Common Venue     7th Most Common Venue  \
21        Medical Center  Mediterranean Restaurant               Men's Store   
23        Massage Studio            Medical Center  Mediterranean Restaurant   
25        Massage Studio            Medical Center  Mediterranean Restaurant   
30                 Motel            Massage Studio            Medical Center   
40         Movie Theater            Medical Center  Mediterranean Restaurant   
44   Monument / Landmark            Massage Studio            Medical Center   
48                 Motel            Massage Studio            Medical Center   
50        Massage Studio            Medical Center  Mediterranean Restaurant   
74                Museum  Mediterranean Restaurant               Men's Store   
90         Movie Theater            Medical Center  Mediterranean Restaurant   

       8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
21             Metro Station    Mexican Restaurant  Middle Eastern Restaurant  
23               Men's Store         Metro Station         Mexican Restaurant  
25               Men's Store         Metro Station         Mexican Restaurant  
30  Mediterranean Restaurant           Men's Store              Metro Station  
40               Men's Store         Metro Station         Mexican Restaurant  
44  Mediterranean Restaurant           Men's Store              Metro Station  
48  Mediterranean Restaurant           Men's Store              Metro Station  
50               Men's Store         Metro Station         Mexican Restaurant  
74             Metro Station    Mexican Restaurant  Middle Eastern Restaurant  
90               Men's Store         Metro Station         Mexican Restaurant

Cluster 4

In [54]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0  Scarborough             3.0  Fast Food Restaurant           Yoga Studio   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                 Motel        Massage Studio        Medical Center   

      6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0  Mediterranean Restaurant           Men's Store         Metro Station   

  9th Most Common Venue     10th Most Common Venue  
0    Mexican Restaurant  Middle Eastern Restaurant

Cluster 5

In [55]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
1               Scarborough             4.0  Construction & Landscaping   
2               Scarborough             4.0              Medical Center   
4               Scarborough             4.0         Fried Chicken Joint   
5               Scarborough             4.0                  Playground   
7               Scarborough             4.0                      Bakery   
9               Scarborough             4.0             College Stadium   
10              Scarborough             4.0           Indian Restaurant   
11              Scarborough             4.0                  Smoke Shop   
12              Scarborough             4.0                      Lounge   
14              Scarborough             4.0                        Park   
15              Scarborough             4.0              Discount Store   
17               North York             4.0                        Pool   
19               North York             4.0         Japanese Restaurant   
22               North York             4.0            Ramen Restaurant   
26               North York             4.0        Caribbean Restaurant   
28               North York             4.0                        Bank   
31               North York             4.0                        Park   
32               North York             4.0                Home Service   
36                East York             4.0                        Park   
37             East Toronto             4.0           Health Food Store   
39                East York             4.0              Sandwich Place   
41             East Toronto             4.0            Greek Restaurant   
42             East Toronto             4.0                        Park   
45          Central Toronto             4.0        Gym / Fitness Center   
47          Central Toronto             4.0                Dessert Shop   
51         Downtown Toronto             4.0                  Restaurant   
52         Downtown Toronto             4.0            Ramen Restaurant   
54         Downtown Toronto             4.0                        Café   
55         Downtown Toronto             4.0                        Café   
56         Downtown Toronto             4.0              Farmers Market   
59         Downtown Toronto             4.0                       Plaza   
61         Downtown Toronto             4.0                        Café   
63          Central Toronto             4.0                Home Service   
64          Central Toronto             4.0                        Park   
65          Central Toronto             4.0              Sandwich Place   
66         Downtown Toronto             4.0                        Café   
67         Downtown Toronto             4.0                        Café   
68         Downtown Toronto             4.0              Airport Lounge   
69   Downtown Toronto Stn A             4.0          Seafood Restaurant   
73                     York             4.0                       Field   
75         Downtown Toronto             4.0               Grocery Store   
76             West Toronto             4.0                      Bakery   
77             West Toronto             4.0                         Bar   
79               North York             4.0  Construction & Landscaping   
80                     York             4.0              Sandwich Place   
82             West Toronto             4.0          Mexican Restaurant   
83             West Toronto             4.0              Breakfast Spot   
84             West Toronto             4.0                        Café   
87    East Toronto Business             4.0                 Yoga Studio   
88                Etobicoke             4.0                 Flower Shop   
91                Etobicoke             4.0              Breakfast Spot   
92                Etobicoke             4.0              Hardware Store   
95                Etobicoke             4.0  

all 5 clusters added successfully